eventuate-common-id/src/main/java/io/eventuate/common/id/DatabaseIdGenerator.java

```diff
@@ -0,0 +1,49 @@
+package io.eventuate.common.id;
+
+import java.util.Optional;
+
+public class DatabaseIdGenerator implements IdGenerator {
+
+  public static final long SERVICE_ID_MAX_VALUE = 0x0000ffffffffffffL;
+  public static final long COUNTER_MAX_VALUE = 0xffffL;
+
+  private final long serviceId;
+
+  @Override
+  public boolean databaseIdRequired() {
+    return true;
+  }
+
+  public DatabaseIdGenerator(long serviceId) {
+    this.serviceId = serviceId;
+
+    if (serviceId < 0 || serviceId > SERVICE_ID_MAX_VALUE) {
+      throw new IllegalArgumentException(String.format("service id should be between 0 and %s", SERVICE_ID_MAX_VALUE));
+    }
+  }
+
+  @Override
+  public Int128 genId(Long databaseId) {
+
+    if (databaseId == null) {
+      throw new IllegalArgumentException("database id is required");
+    }
+
+    return new Int128(databaseId, serviceId);
+  }
+
+  @Override
+  public Optional<Int128> incrementIdIfPossible(Int128 anchorId) {
```

Comment:
Since a client wants an Int128 and will call the IdGenerator again if `incrementIdIfPossible()` returns `Optional.empty()`  I think a simpler interface would be:

```
public Int128 genNextId(Optional<Int128> possibleCurrentId, ...) {
   // possibleCurrentId.isEmpty() => genId(...)
   // possibleCurrentId.counter == MAX => genId(...)
  // else incrementCounter
}
```

possibleCurrentId would be empty the first time, subsequent times it would be the value returned by the previous call.